### DISCUSSION TITLE
How to implement a custom scalar to check that type is a date

### DISCUSSION THREAD

**Discussion by wakywayne** [hayes/pothos#650] [Q&A] [answered]
### I have added the code to my schema as instructed in the docs.:

```
    Scalars: {
        Date: {
            Input: Date;
            Output: Date;
        };
    };
```

### I tried doing this from the docs: 
*but obviously that represents a function that I don't know how to make*

```
builder.addScalarType('Date', CustomDateScalar, {});
```

### So now I tried something like this: 

```
builder.scalarType('Date', {
    serialize: (value) => value,
    parseValue: (value) => {
        if(typeof value === typeof Date) {
            return new value
        }else{
            throw new Error('Date scalar can only be serialized from a Date object')
        }
    },

});
```

**I do not have the slightest clue what these serialize and parse value do and I cant find there definitions in the docs**

**Comment by hayes** [comment, marked as answer]
If you don't need custom serialization or parsing for your scalars, I would recommend using an existing scalar instead of defining your own.  The graphql-scalar package has tons of great scalars that cover most use cases.  You can see a fairly simple example of a date scalar added here: https://github.com/hayes/pothos/blob/main/examples/complex-app/src/builder.ts


parsing and serialization are required for converting between JSON (in the request  or response) and the values used in your resolvers.   For example, with the date scalar, the argument passed to parse would be a iso date string, and the function should return a Date object.  Serialization is the opposite, it receives a Date object, and needs to turn it in to a iso date string so it can be accurately represented in the JSON response.  Generally I would recommend using the existing implementations so you don't have to define these methods yourself, unless you need something very custom

**Reply by wakywayne** [reply]
Thank you so much for taking the time to answer my question I really appreciate it. I also need to make a type for a MongoDB objectID. I was importing the type from BSON and trying to use that directly is this possible or no? I think my understanding of the type field is infantile, I am using it as if it is just a typescript field. Do you think you can explain how the type field actually works?

**Reply by hayes** [reply]
Not near a computer, and not very familiar with mongo.

The `type` option on a field determines the GraphQL type for the field.  You can pass in a string with the name of a Scalar type, or a reference to an object, union, or interface type you defined with the builder.   When you use methods like builder.objectType or builder.objectRef the returned value is a reference to the corresponding type, which you can pass to type option of a field.

To define a field based on your object ID you first need to figure out the appropriate graphql type.   This might just be an ID if it can be easily converted to/from a string.  It's also possible you would want to define a custom scalar for these values, but that's probably exposing implementation details in your API in a way I wouldn't recommend.

**Reply by wakywayne** [reply]
Thank you so much I was really screwed until you helped me my understanding of type was likely going to put me down a rabbit hole
